Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselPowerSupply.cs

```diff
@@ -0,0 +1,334 @@
+                {
+                    AbstractScript = new DefaultDieselPowerSupply();
+                }
+
+                AssignScriptFunctions();
+
+                Script.Initialize();
+                Activated = true;
+            }
+
+            TractionCutOffRelay.Initialize();
+        }
+
+
+        //================================================================================================//
+        /// <summary>
+        /// Initialization when simulation starts with moving train
+        /// <\summary>
+        public override void InitializeMoving()
+        {
+            base.InitializeMoving();
+
+            TractionCutOffRelay.InitializeMoving();
+        }
+
+        public override void Save(BinaryWriter outf)
+        {
+            outf.Write(ScriptName);
+
+            base.Save(outf);
+            TractionCutOffRelay.Save(outf);
+        }
+
+        public override void Restore(BinaryReader inf)
+        {
+            ScriptName = inf.ReadString();
+
+            base.Restore(inf);
+            TractionCutOffRelay.Restore(inf);
+        }
+
+        public override void Update(float elapsedClockSeconds)
+        {
+            base.Update(elapsedClockSeconds);
+
+            TractionCutOffRelay.Update(elapsedClockSeconds);
+
+            Script?.Update(elapsedClockSeconds);
+        }
+
+        protected override void AssignScriptFunctions()
+        {
+            base.AssignScriptFunctions();
+
+            // DieselPowerSupply getters
+            Script.CurrentDieselEnginesState = () => DieselLocomotive.DieselEngines.State;
+            Script.CurrentDieselEngineState = (id) =>
+            {
+                if (id >= 0 && id < DieselEngines.Count)
+                {
+                    return DieselEngines[id].State;
+                }
+                else
+                {
+                    return DieselEngineState.Unavailable;
+                }
+            };
+            Script.CurrentTractionCutOffRelayState = () => TractionCutOffRelay.State;
+            Script.TractionCutOffRelayDriverClosingOrder = () => TractionCutOffRelay.DriverClosingOrder;
+            Script.TractionCutOffRelayDriverOpeningOrder = () => TractionCutOffRelay.DriverOpeningOrder;
+            Script.TractionCutOffRelayDriverClosingAuthorization = () => TractionCutOffRelay.DriverClosingAuthorization;
+
+            // DieselPowerSupply setters
+            Script.SignalEventToDieselEngines = (evt) => DieselEngines.HandleEvent(evt);
+            Script.SignalEventToDieselEngine = (evt, id) => DieselEngines.HandleEvent(evt, id);
+            Script.SignalEventToTractionCutOffRelay = (evt) => TractionCutOffRelay.HandleEvent(evt);
+        }
+    }
+
+    public class DefaultDieselPowerSupply : DieselPowerSupply
+    {
+        private Timer PowerOnTimer;
+        private Timer AuxPowerOnTimer;
+
+        private DieselEngineState PreviousFirstEngineState;
+        private DieselEngineState PreviousSecondEngineState;
```

Comment:
Why should we be limited to two diesel engines? Perhaps a dictionary would be the way to go here.

Reply:
We are only using these variables for the EnginePowerOn/Off and SecondEnginePowerOn/Off events.